contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,773 @@
+    Product memory product = _products[params.productId];
+    require(product.initialPriceRatio != 0, "Cover: Product not initialized");
+    require(
+      isAssetSupported(product.coverAssets, params.payoutAsset),
+      "Cover: Payout asset is not supported"
+    );
+    require(params.period >= MIN_COVER_PERIOD, "Cover: Cover period is too short");
+    require(params.period <= MAX_COVER_PERIOD, "Cover: Cover period is too long");
+    require(params.commissionRatio <= MAX_COMMISSION_RATIO, "Cover: Commission rate is too high");
+
+    uint totalPremiumInNXM = _buyCover(params, _coverData.length, allocationRequests);
+
+    IPool _pool = pool();
+    uint tokenPriceInPaymentAsset = _pool.getTokenPrice(params.paymentAsset);
+    (, uint8 paymentAssetDecimals) = _pool.coverAssets(params.paymentAsset);
+
+    uint premiumInPaymentAsset = totalPremiumInNXM * (tokenPriceInPaymentAsset / 10 ** paymentAssetDecimals);
+    require(premiumInPaymentAsset <= params.maxPremiumInAsset, "Cover: Price exceeds maxPremiumInAsset");
+
+    if (params.payWithNXM) {
+      retrieveNXMPayment(totalPremiumInNXM, params.commissionRatio, params.commissionDestination);
+    } else {
+      retrievePayment(
+        premiumInPaymentAsset,
+        params.paymentAsset,
+        params.commissionRatio,
+        params.commissionDestination
+      );
+    }
+
+    // push the newly created cover
+    _coverData.push(CoverData(
+        params.productId,
+        params.payoutAsset,
+        0 // amountPaidOut
+      ));
+
+    uint coverId = _coverData.length - 1;
+    ICoverNFT(coverNFT).safeMint(params.owner, coverId);
+
+    // Enable this when cover amount tracking is necessary
+    if (coverAmountTrackingEnabled) {
+      totalActiveCoverInAsset[params.payoutAsset] += params.amount;
+    }
+
+    emit CoverBought(coverId, params.productId, 0, msg.sender, params.ipfsData);
+    return coverId;
+  }
+
+  function _buyCover(
+    BuyCoverParams memory params,
+    uint coverId,
+    PoolAllocationRequest[] memory allocationRequests
+  ) internal returns (uint totalPremiumInNXM) {
+
+    // convert to NXM amount
+    uint nxmPriceInPayoutAsset = pool().getTokenPrice(params.payoutAsset);
+    uint remainderAmountInNXM = 0;
+    uint totalCoverAmountInNXM = 0;
+
+    uint _coverSegmentsCount = _coverSegments[coverId].length;
+
+    for (uint i = 0; i < allocationRequests.length; i++) {
+
+      uint requestedCoverAmountInNXM
+        = allocationRequests[i].coverAmountInAsset * 1e18 / nxmPriceInPayoutAsset + remainderAmountInNXM;
+
+      (uint coveredAmountInNXM, uint premiumInNXM, uint rewardsInNXM) = allocateCapacity(
+        params,
+        coverId,
+        stakingPool(allocationRequests[i].poolId),
+        requestedCoverAmountInNXM
+      );
+
+      // apply the global rewards ratio and the total Rewards in NXM
+      tokenController().mintStakingPoolNXMRewards(rewardsInNXM, allocationRequests[i].poolId);
```

Comment:
Let's cache `tokenController` in a variable before the loop (there's space).

Reply:
leaving for post-v2